service/r2/service.go

```diff
@@ -553,3 +565,39 @@ type ruleSetJSON struct {
 	MappingRules  []mappingRuleJSON `json:"mappingRules"`
 	RollupRules   []rollupRuleJSON  `json:"rollupRules"`
 }
+
+// Creates a new RuleSetSnapshot from a rulesetJSON. If the ruleSetJSON has no IDs for any of its
+// mapping rules or rollup rules, it generates missing IDs and sets as a string UUID string so they
+// can be stored in a mapping (id -> rule).
+// This is used in the case where a new rollupRule or mappingRule has been created from the client
+// side and thusly does not have a ID set for it.
+func (r ruleSetJSON) ruleSetSnapshot(opts ruleSetSnapshotOpts) *rules.RuleSetSnapshot {
+	rss := rules.RuleSetSnapshot{
+		Namespace:    r.Namespace,
+		Version:      r.Version,
+		MappingRules: map[string]*rules.MappingRuleView{},
+		RollupRules:  map[string]*rules.RollupRuleView{},
+	}
+
+	for _, mr := range r.MappingRules {
+		id := mr.ID
+		if id == "" && opts.generateMissingID {
+			id = uuid.New()
+			mr.ID = id
+		}
+		rss.MappingRules[id] = mr.mappingRuleView()
```

Comment:
If the mapping id is an empty string and opts.generateMissingID is false this will continuously overwrite `rss.MappingRules[""]`. Im not sure what the best behavior is, maybe just leave it out of the list? But I think just taking whatever the last rule is and adding that to the snapshot while dropping all the other ones is kind of confusing.

Reply:
Updated per in person conversation